File: graph/src/547_number_of_provinces.c
#include <stdlib.h>

void checkCycle(int idx, int n, int** isConnected, int* visited){
    visited[idx] = 1;
    for(int i = 0; i < n; i++){
        // std::cout<<"idx: "<<idx<<"adj "<<i<<"connected status: "<<isConnected[idx][i]<<"visited status: "<<visited[i]<<std::endl;
        // if(i == idx)continue;
        if(visited[i]||!isConnected[idx][i])continue;
        checkCycle(i, n, isConnected, visited);
    }
}

int findCircleNum(int** isConnected, int isConnectedSize, int* isConnectedColSize){
    int res = 0;
    for(int i = 0; i < isConnectedSize; i++){
        if(isConnectedColSize[i])continue;
        checkCycle(i, isConnectedSize, isConnected, isConnectedColSize);
        res++;
    }
    return res;
}

File: graph/src/547_number_of_provinces.py

def checkCycle(idx, n, isConnected, visited):
    visited[idx] = 1
    for i in range(n):
        if(visited[i] or not(isConnected[idx][i])):
            continue
        checkCycle(i, n, isConnected, visited)

def findCircleNum(isConnected):
    res = 0
    numCities = len(isConnected)
    visited = [0] * numCities
    for i in range(numCities):
        if visited[i]:
            continue
        checkCycle(i, numCities, isConnected, visited)
        res += 1
    
    return res
